### ISSUE TITLE
annotation seems to be missing data

### ISSUE THREAD

**Issue by yasirs** [jokergoo/ComplexHeatmap#142]
The following toy data set seems to exemplify the problem:

```
set.seed(0)
mm = matrix(data = rnorm(1000 * 8), nrow = 8)
rownames(mm) = letters[1:8]
annodf = data.frame(var1=rbinom(1000, 1, .01))

library(ComplexHeatmap)
ha = HeatmapAnnotation(df=annodf, col=list("var1"=c("0"="white","1"="red")))
ht = Heatmap(mm, show_column_dend = F, 
        show_column_names = F, bottom_annotation = ha)
draw(ht)
```

 I can make out 9 distinct lines in the annotation, though one of them is thicker and may be composed of multiple lines.
![heatplot](https://user-images.githubusercontent.com/400538/34901939-a60e4428-f7df-11e7-915d-3af2b8f56bdf.png)

I can also plot the annotation separately as a barplot to see what I think it should look like as
`barplot(annodf$var1[column_order(ht)])
`
This gives me:  
![barplot](https://user-images.githubusercontent.com/400538/34901940-ad24dd1c-f7df-11e7-99ab-9c0b1b7ef65a.png)

This looks like 11 lines, with one probably composed of two. We can look at the data:

```
> sum(annodf$var1)
[1] 12
> which(annodf$var1[column_order(ht)]==1)
 [1] 112 126 128 250 430 477 561 568 643 673 893 934
```
to confirm that the barplot is correct.

Why does the annotation seem to be missing data?

**Update:**

I don't seem to see this discrepancy in smaller datasets. To illustrate this point, and to confirm that the method of writing the barplot of the annotation is correct, I do so for a 5x5 plot:

    set.seed(0)
    mm2 = matrix(data = rnorm(5 * 5), nrow = 5)
    rownames(mm2) = letters[1:5]
    colnames(mm2) = LETTERS[1:5]
    annodf2 = data.frame(var1=rbinom(5, 1, .45))
    ha2 = HeatmapAnnotation(df=annodf2, col=list("var1"=c("0"="white","1"="red")))
    ht2 = Heatmap(mm2, show_column_dend = F, 
                 show_column_names = F, bottom_annotation = ha2)
    draw(ht2)
    barplot(annodf2$var1[column_order(ht2)])

The heatmap annotation:[![enter image description here][3]][3], and the barplot:[![enter image description here][4]][4]

These match, as the 2nd and the 5th columns have var1=1. So why don't the annotation and the barplot match for the larger plot?

  [3]: https://i.stack.imgur.com/Mn5B7.png
  [4]: https://i.stack.imgur.com/i03f6.png

Crossposted: https://stackoverflow.com/questions/48236408/complexheatmap-annotation-seems-to-be-missing-data

**Comment by jokergoo** [comment]
How did you save the plot? Did you directly save it as png or pdf or you just made a screenshot?

Following is what I got from my laptop.

<img width="1257" alt="screen shot 2018-01-18 at 15 01 03" src="https://user-images.githubusercontent.com/449218/35101609-7203a882-fc60-11e7-935c-5d8815372f87.png">

The reason why you found one red line missing in the heatmap might be that it was overwritten by neignbouring white lines. When you specify colors as `c("0"="white","1"="red")`, it actually also draw vertical lines for values of 0. With different graphic devices (e.g. png or pdf), you will probably see the hidden red line.

In order not draw anything for the 0 values, you can set colors by `c("0"=NA,"1"="red")` or `c("0"="#FFFFFF00","1"="red")` (to set complete transparent color for 0).

**Comment by jokergoo** [comment]
Can you also attach a simple code?

I use `pdf()` with `use_raster = TRUE` in my daily analysis and it works well.

**Comment by dyusuf** [comment]
I tried to reproduce the issue with the example from yasirs, it works well as you said. interesting.

but with my own data, the issue indeed persists.  I will send you the rds via email.

```
library(ComplexHeatmap)

index_ <- readRDS('index_.rds')
tags <- readRDS('tags.rds')
yes_no <- readRDS('yes_no.rds')

ha_row <- rowAnnotation(link = row_anno_link(at = index_, side = "left", labels = tags, labels_gp = gpar(fontsize = 5), link_width = unit(5, "mm")), width = max_text_width(labels, gp = gpar(fontsize = 5)) + unit(5, "mm"))

ht_T <- Heatmap(yes_no, name='transcription factor', width = unit(2, "mm"), column_names_gp = gpar(fontsize = 8), heatmap_legend_param = list(title_gp = gpar(fontsize = 6, fontface = "bold"), labels_gp = gpar(fontsize = 5)), col= c(no="#FFFFFF00", yes="#800000"), show_row_names=FALSE, show_column_names=FALSE, use_raster = TRUE, raster_quality = 2)


pdf('use_raster_T.pdf')
draw(ht_T + ha_row)
dev.off()

ht_F <- Heatmap(yes_no, name='transcription factor', width = unit(2, "mm"), column_names_gp = gpar(fontsize = 8), heatmap_legend_param = list(title_gp = gpar(fontsize = 6, fontface = "bold"), labels_gp = gpar(fontsize = 5)), col= c(no="#FFFFFF00", yes="#800000"), show_row_names=FALSE, show_column_names=FALSE, use_raster = FALSE, raster_quality = 2)

pdf('use_raster_F.pdf')
draw(ht_F + ha_row)
dev.off()

```

[use_raster_F.pdf](https://github.com/jokergoo/ComplexHeatmap/files/2206625/use_raster_F.pdf)
[use_raster_T.pdf](https://github.com/jokergoo/ComplexHeatmap/files/2206626/use_raster_T.pdf)

**Comment by dyusuf** [comment]
@jokergoo any feedback? have I made any mistakes, or there is indeed an issue?

**Comment by jokergoo** [comment]
Sorry for the late reply! I have tested your code with the three cases, and both work fine without any error.

Can you try with the newest version of ComplexHeatmap? Or can you send me the code which produces error? The code you sent me before runs successfully, right? I need to see the error message.

**Comment by dyusuf** [comment]
I am using bioconductor-complexheatmap 1.17.1 from bioconda.

I could try out the latest one 1.18.1 from bioconductor.

the code I sent to you showed the effect of `use_raster` on the PDF output, however, there were no errors when executing the codes. the differences are in the PDF outputs.  Sorry that it was not clear.

`ht_T` was generated with  `use_raster = TRUE` and the corresponding pdf is `use_raster_T.pdf`

`ht_F` was generated with  `use_raster = FALSE` and the corresponding pdf is `use_raster_F.pdf`

I attached the two pdfs in the previous post.

I use term `tag` to show where color should appear. In `use_raster_F.pdf`,  the colors appear at all the tags while it is not the case in `use_raster_T.pdf` e.g. the last two tags at the bottom.

**Comment by jokergoo** [comment]
Ah, I finally know what you mean. Sorry that I didn't read very carefully to your post!

I modified your code a little bit that I just add all heatmaps/annotations into one heatmap list:

```
index_ <- readRDS('index_.rds')
tags <- readRDS('tags.rds')
yes_no <- readRDS('yes_no.rds')

ha_row <- rowAnnotation(link = row_anno_link(at = index_, side = "left", labels = tags, 
	labels_gp = gpar(fontsize = 5), link_width = unit(5, "mm")), width = max_text_width(labels, gp = gpar(fontsize = 5)) + unit(5, "mm"))

ht_T <- Heatmap(yes_no, name='transcription factor', width = unit(2, "mm"), 
	column_names_gp = gpar(fontsize = 8), heatmap_legend_param = list(title_gp = gpar(fontsize = 6, fontface = "bold"), 
		labels_gp = gpar(fontsize = 5)), col= c(no="#FFFFFF00", yes="#800000"), 
	show_row_names=FALSE, show_column_names=FALSE, use_raster = TRUE, raster_quality = 2,
	column_title = "use_raster = T")

ht_list_1 = ht_T + ha_row

ht_F <- Heatmap(yes_no, name='transcription factor', width = unit(2, "mm"), 
	column_names_gp = gpar(fontsize = 8), heatmap_legend_param = list(title_gp = gpar(fontsize = 6, fontface = "bold"), 
		labels_gp = gpar(fontsize = 5)), col= c(no="#FFFFFF00", yes="#800000"), 
	show_row_names=FALSE, show_column_names=FALSE, use_raster = FALSE, raster_quality = 2,
	column_title = "use_raster = FALSE")


ht_list_2 = ht_F + ha_row

pdf("test.pdf", width = 10, height = 6)
draw(ht_list_1 + ht_list_2, gap = unit(c(2, 30, 2), "mm"))
dev.off()

png("test.png", width = 1000, height = 600)
draw(ht_list_1 + ht_list_2, gap = unit(c(2, 30, 2), "mm"))
dev.off()
```

And the two figures that I generated with above code:

[test.pdf](https://github.com/jokergoo/ComplexHeatmap/files/2218859/test.pdf)
[test.png](https://user-images.githubusercontent.com/449218/43070659-b8dc18b4-8e70-11e8-8bfc-175dbf74f8c2.png)

Basically they are fine. The "yes" lines are at where they should locate, no matter it is pdf or png. I would suspect might it be that your computer does not support transparency for png figures? Can you set `NA` as "no" colors? `NA` basically means do not draw anything.

**Comment by dyusuf** [comment]
thanks for the feedback!

I set `col= c(no=NA, yes="#800000")` for the new testing.

there still is the issue with `use_raster = T` in the outputs regardless of pdf or png. 

[test.png](https://user-images.githubusercontent.com/11738833/43074387-5bc63350-8e7d-11e8-974a-f90396ff431d.png)
[test.pdf](https://github.com/jokergoo/ComplexHeatmap/files/2219113/test.pdf)

yes, I agree with this more likely related to my system, might be due to the conda packages.

could you send me your sessionInfo()?

**Comment by jokergoo** [comment]
This is so strange to look at your plots. Can you add argument `raster_device = "CairoPNG"` to use CairoPNG as the png device (you need to install **Cairo** package)? or what if use `jpeg` as device?

Following is my session info:

```
> sessionInfo()
R version 3.4.4 (2018-03-15)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.2

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
[1] ComplexHeatmap_1.17.1 colorout_1.2-0

loaded via a namespace (and not attached):
[1] colorspace_1.3-2    compiler_3.4.4      rjson_0.2.20
[4] circlize_0.4.5      GlobalOptions_0.1.1 RColorBrewer_1.1-2
[7] GetoptLong_0.1.7    shape_1.4.4         png_0.1-7
```

**Comment by dyusuf** [comment]
yes, with `raster_device = "CairoPNG"` the output looks better at least for pdf. :)

thanks a lot!

[test_new.jpg](https://user-images.githubusercontent.com/11738833/43077155-f0de603a-8e86-11e8-9601-bc9f928aaad8.jpg)
[test_new.png](https://user-images.githubusercontent.com/11738833/43077156-f0f6f866-8e86-11e8-9c00-35839118ab51.png)
[test_new.pdf](https://github.com/jokergoo/ComplexHeatmap/files/2219303/test_new.pdf)
